### DISCUSSION TITLE
v7 cannot play a certain clip accurately

### DISCUSSION THREAD

**Discussion by xinyuaixiaoyu** [katspaugh/wavesurfer.js#3310] [Q&A]
Through the region-out callback, there will be a time error in pausing playback, which is probably a few tenths of a second.

**Comment by katspaugh** [comment]
We use requestAnimationFrame for polling the current time position. It fires approximately every 16 ms, so that would be the accuracy of region-out.

**Reply by xinyuaixiaoyu** [reply]
So how to accurately play the audio of a region or a certain time period?

**Reply by xinyuaixiaoyu** [reply]
In the phonetic scene model, asynchronous callbacks can cause accuracy problems

**Reply by katspaugh** [reply]
You could use a Web Audio buffer for this, it supports precise scheduling of playback.

Something like this:

```js
import WebAudioPlayer from 'https://unpkg.com/wavesurfer.js@7/dist/webaudio.js'

WebAudioPlayer.prototype.stopAt = (timeSeconds) => {
  const delay = timeSeconds - this.currentTime
  this.bufferNode?.stop(this.audioContext.currentTime + delay)

  this.bufferNode?.addEventListener('ended', () => {
    this.bufferNode = null
    this.pause()
  }, { once: true })
}

const audio = new WebAudioPlayer()

const wavesurfer = new WaveSurfer({
  container: '#...',
  url: 'https://...', // audio URL
  media: audio, // pass the web audio player
})

regions.on('region-in', (region) => {
  audio.stopAt(region.end)
})
```

**Reply by xinyuaixiaoyu** [reply]
What version is the audio parameter? @7.4.2 cannot find this parameter

**Reply by xinyuaixiaoyu** [reply]
And ts+react will report an error when adding a method to the prototype. After ignoring it, the method execution cannot find this.

**Reply by katspaugh** [reply]
Oh, sorry, it's `media`, not `audio`.

**Reply by xinyuaixiaoyu** [reply]
In the ts+react version, an error will be reported when adding the prototype, and the private property bufferNode cannot be read. Is there any react case?

**Reply by katspaugh** [reply]
Feel free to make a PR to add that method

**Reply by xinyuaixiaoyu** [reply]
The PR has been added according to the code you gave.
https://github.com/katspaugh/wavesurfer.js/pull/3325

**Reply by xinyuaixiaoyu** [reply]
Setting the media attribute to WebAudioPlayer will also report an error and the type does not match.

**Reply by katspaugh** [reply]
I know, you have to cast it. I might fix it in future.

**Reply by xinyuaixiaoyu** [reply]
When is this function expected to be launched? It has a great impact on current use.

**Reply by katspaugh** [reply]
Released it in 7.4.6.

**Reply by xinyuaixiaoyu** [reply]
Updated version 7.5.2, media still has ts type errors for WebAudio instances, and the playback will freeze and drop frames, causing the browser to freeze.
